lua/neo-tree/git/ignored.lua

```diff
@@ -83,26 +74,77 @@ M.mark_ignored = function(state, items)
         end
       end
     end
+    return result
+  end
 
-    vim.list_extend(all_results, result)
+  local function finalize(all_results)
+    local show_anyway = state.filtered_items and state.filtered_items.hide_gitignored == false
+    log.trace("IGNORED: Comparing results to mark items as ignored, show_anyway:", show_anyway)
+    local ignored, not_ignored = 0, 0
+    for _, item in ipairs(items) do
+      if M.is_ignored(all_results, item.path, item.type) then
+        item.filtered_by = item.filtered_by or {}
+        item.filtered_by.gitignored = true
+        item.filtered_by.show_anyway = show_anyway
+        ignored = ignored + 1
+      else
+        not_ignored = not_ignored + 1
+      end
+    end
+    log.trace("IGNORED: mark_ignored is complete, ignored:", ignored, ", not ignored:", not_ignored)
+    log.trace("================================================================================")
   end
 
-  local show_anyway = state.filtered_items and state.filtered_items.hide_gitignored == false
-  log.trace("IGNORED: Comparing results to mark items as ignored, show_anyway:", show_anyway)
-  local ignored, not_ignored = 0, 0
-  for _, item in ipairs(items) do
-    if M.is_ignored(all_results, item.path, item.type) then
-      item.filtered_by = item.filtered_by or {}
-      item.filtered_by.gitignored = true
-      item.filtered_by.show_anyway = show_anyway
-      ignored = ignored + 1
-    else
-      not_ignored = not_ignored + 1
+  local all_results = {}
+  if type(callback) == "function" then
+    local jobs = {}
+    local progress = 0
+    for folder, folder_items in pairs(folders) do
+      local args = { "-C", folder, "check-ignore", "--stdin" }
+      local job = Job:new({
+        command = "git",
+        args = args,
+        enabled_recording = true,
+        writer = folder_items,
+        on_start = function()
+          log.trace("IGNORED: Running async git with args: ", args)
+        end,
+        on_exit = function(self, code, _)
+          local result
+          if code ~= 0 then
+            log.debug("Failed to load ignored files for", state.path, ":", self:stderr_result())
+            result = {}
+          else
+            result = self:result()
+          end
+          vim.list_extend(all_results, process_result(result))
+          progress = progress + 1
+          if progress == #jobs then
```

Comment:
I'm always on the hunt for unexpected memory leaks. In this case, the garbage collector _should_ clean this up, but just in case, we don't really need to keep a reference to these jobs at all. The number of jobs is equal to the number of folders, so you can just compare that instead.
```suggestion
          if progress == #folders then
```

Then you can just go ahead and start the job as soon as it is defined without storing any reference to it at all. i.e. 
```lua
Job:new({ 
  ...
}):start()
```

Reply:
Ah yes, you are absolutely correct. I was thinking that folders was a list for some reason.